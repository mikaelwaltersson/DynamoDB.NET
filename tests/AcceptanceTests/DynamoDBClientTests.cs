using Amazon.DynamoDBv2;
using DynamoDB.Net;
using DynamoDB.Net.Model;
using DynamoDB.Net.Serialization;
using Microsoft.Extensions.DependencyInjection;
using Microsoft.Extensions.Options;
using Snapshooter.Xunit;
using Xunit;

namespace DynamoDB.NET.Tests.AcceptanceTests
{
    public partial class DynamoDBClientTests
    {
        [Fact]
        public async Task CanCreateTables()
        {
 
            var services = new ServiceCollection();

            services.AddOptions();
            services.AddLogging();
            
            services.AddSingleton(
                typeof(IAmazonDynamoDB),
                serviceProvider =>
                    new AmazonDynamoDBClient(
                        new AmazonDynamoDBConfig
                        {
                            ServiceURL = "http://localhost:8000"
                        }));

            services.AddSingleton<IDynamoDBClient, DynamoDBClient>();
            services.Configure<DynamoDBClientOptions>(
                options => 
                {
                    options.TableNamePrefix = "acceptance-tests-";
                }
            );

            var serviceProvider = services.BuildServiceProvider();

            var dynamoDb = serviceProvider.GetRequiredService<IAmazonDynamoDB>();
            var dynamoDbClientOptions = serviceProvider.GetRequiredService<IOptions<DynamoDBClientOptions>>();
            var client = serviceProvider.GetRequiredService<IDynamoDBClient>();

            var table = TableDescription.Get(typeof(TestModels.UserPost), JsonContractResolver.DefaultDynamoDB);
            var createTableRequest = table.GetCreateTableRequest(dynamoDbClientOptions.Value);
            
            var result = await dynamoDb.CreateTableAsync(createTableRequest);
            try 
            {
                Assert.Equal("acceptance-tests-user-posts", result.TableDescription.TableName);

                var item1 = await client.PutAsync(
                    new TestModels.UserPost 
                    { 
                        UserId = new Guid("00000000-0000-0000-0000-000000000001"), 
                        Timestamp = new DateTime(2022, 10, 18, 16, 42, 0),
                        RoleIds =
                        {
                            new Guid("00000000-0000-0000-0000-000000000002"), 
                            new Guid("00000000-0000-0000-0000-000000000003")
                        }
                    });
                var item2 = await client.TryGetAsync(new PrimaryKey<TestModels.UserPost>(new Guid("00000000-0000-0000-0000-000000000001"), new DateTime(2022, 10, 18, 16, 42, 0)));

                Assert.NotNull(item2);
                Snapshot.Match(item1);
            }
            finally 
            {
                await dynamoDb.DeleteTableAsync(result.TableDescription.TableName);
            }
        }
    }
}