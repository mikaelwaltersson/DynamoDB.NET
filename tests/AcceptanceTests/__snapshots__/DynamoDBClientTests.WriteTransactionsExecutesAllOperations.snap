[
  {
    "userId": {
      "S": "40822a39-1d07-4170-9486-000000000004"
    },
    "timestamp": {
      "S": "2022-10-18T16:04:00.0000000+00:00"
    },
    "roleIds": {
      "L": [
        {
          "S": "4bd6e5ff-3f77-4895-b861-000000000004"
        },
        {
          "S": "34f9d49a-706d-4bed-b9e8-000000000004"
        }
      ]
    }
  },
  {
    "userId": {
      "S": "40822a39-1d07-4170-9486-000000000003"
    },
    "timestamp": {
      "S": "2022-10-18T16:03:00.0000000+00:00"
    },
    "roleIds": {
      "L": [
        {
          "S": "4bd6e5ff-3f77-4895-b861-000000000003"
        },
        {
          "S": "34f9d49a-706d-4bed-b9e8-000000000003"
        }
      ]
    }
  },
  {
    "userId": {
      "S": "40822a39-1d07-4170-9486-000000000001"
    },
    "roleIds": {
      "L": [
        {
          "S": "4bd6e5ff-3f77-4895-b861-000000000001"
        },
        {
          "S": "34f9d49a-706d-4bed-b9e8-000000000001"
        },
        {
          "S": "12d73fb1-5286-4376-b3b5-e03eb6a4ec13"
        },
        {
          "S": "8f671114-c522-4a5b-9490-0165a1b8e56e"
        }
      ]
    },
    "timestamp": {
      "S": "2022-10-18T16:01:00.0000000+00:00"
    }
  }
]
